napari/plugins/hook_specifications.py

```diff
@@ -317,3 +328,53 @@ def napari_write_vectors(path: str, data: Any, meta: dict) -> Optional[str]:
         If data is successfully written, return the ``path`` that was written.
         Otherwise, if nothing was done, return ``None``.
     """
+
+
+# -------------------------------------------------------------------------- #
+#                                 GUI Hooks                                  #
+# -------------------------------------------------------------------------- #
+
+
+@napari_hook_specification(historic=True)
+def napari_experimental_provide_functions() -> Union[
+    Tuple[Callable, Dict], Sequence[Tuple[Callable, Dict]]
+]:
+    """Provide functions and args that can be passed to magicgui.
+
+    Returns
+    -------
+    functions : tuple of callable, dict or sequence tuple of callable, dict
+        Tuple of 2-tuple, where each tuple has a callable function a dictionary of
+        keyword arguments for magicgui. The following class attributes may be used
+        to control how how the plugin appears in the menu.
+            napari_menu_name (str): If provided, the name that will appear in the
+                plugin menu. Otherwise function name will be used.
+            napari_area (str): The initial area (relative to the viewer window)
+                to show this widget.  Must be one of "top", "left", "bottom",
+                "right"
+            napari_allowed_areas (List[str]): The areas (relative to the viewer
+                window) where this widget can be moved. List members must be
+                one of "top", "left", "bottom", "right".
+    """
+
+
+@napari_hook_specification(historic=True)
+def napari_experimental_provide_dock_widgets() -> Union[
+    Type[QWidget], Sequence[Type[QWidget]]
```

Comment:
@tlambert03 we're getting fails from our headless mode check because of the qtpy imports. Should I change this to use strings and be forward reference, or do import inside `if TYPE_CHECKING`?

Reply:
yep (both)!  import inside `if typing.TYPE_CHECKING` ... and then annotated with `Type['QWidget']`

Reply:
(it will pass the checks with the forward reference alone, but mypy wouldn't work without the import inside of TYPE_CHECKING)

Reply:
ok, thanks I will give this a try now!

Reply:
ok, change pushed. let me know if that looks good